Wip

CoitThomas opened the pull request.

CoitThomas committed:
First draft of make_european.py with data_input.txt file.

CoitThomas committed:
Add test modules.
Add more rigorous assertions.
Combine make_european function with convert_data function.

austinwiltshire commented on make_european.py:
```diff
@@ -0,0 +1,82 @@
+    European format and convert the username to all lowercase. Return
+    the new date and username.
+    """
+    assert isinstance(string, str), "Initial input for convert_input needs to be a string."
```
We worked on these for awhile so you could get the hang of asserts, but actually type checking like this in python or ruby is considered 'unpythonic'. For instance, there might be other objects that are 'string-like' enough for your code to work on that these assertions would bar.

Are there other assertions you can add, either at the top here or in your code, that asserts more on the _formatting_ of the string, for instance?

austinwiltshire commented on make_european.py:
```diff
@@ -0,0 +1,82 @@
+
+    username, host = find_email_address(string)
+    if host.lower() == desired_host:
+        month, day, year = find_date(string)
```
this is very well factored!

austinwiltshire commented on make_european.py:
```diff
@@ -0,0 +1,82 @@
+    assert isinstance(string, str), "The input for find_email_address needs to be a string."
+
+    email_address = re.search(r'([\w.-]+)@([\w.-]+)', string)
+    assert email_address, "A proper email address must be provided."
```
this is a great assertion.

also, check out: http://www.regular-expressions.info/email.html

i'm going to judge your own code by it passing the tests - its a great example of YAGNI. none of the emails i provided beat your regex, so that regex is good enough for our requirements. BUT, matching an email is notoriously hard, so go ahead and read that for some surrounding thoughts.

austinwiltshire commented on make_european.py:
```diff
@@ -0,0 +1,82 @@
+    email_address = re.search(r'([\w.-]+)@([\w.-]+)', string)
+    assert email_address, "A proper email address must be provided."
+
+    username = email_address.group(1)
```
https://docs.python.org/2/library/re.html

search for 'named groups' in that page and use those instead of the numbers. this increases regex readability.

austinwiltshire commented on make_european.py:
```diff
@@ -0,0 +1,82 @@
+    assert isinstance(string, str), "The function find_date takes a string."
+
+    date = re.search('([0-9]{1,2})/([0-9]{1,2})/([0-9]{4})', string)
+    assert date, "A date in the format mm/dd/yyyy must be provided."
```
this is a good assertion

austinwiltshire commented on make_european.py:
```diff
@@ -0,0 +1,82 @@
+    found in the string."""
+    assert isinstance(string, str), "The function find_date takes a string."
+
+    date = re.search('([0-9]{1,2})/([0-9]{1,2})/([0-9]{4})', string)
```
playing around with whitespacing can make code, especially terse stuff like regex's, much more readable. i _think_ this formatting will work but you may have to toy with it - change this and other regexes to use the following formatting (i'll give this one as an example)

```
date = re.search(
    "([0-9]{1,2})" #the year
    "/([0-9]{1,2})" #the month
    "/([0-9]{4})", #the day
    string
)

if you combine the above with named groups, the comments become redundant. but there are times when something has semantic meaning but you don't want to capture it (so don't need a named group) and comments can be helpful there. i *believe* python auto concatenates all strings that are next to each other so it turns "hello " "world" into "hello world", even over line breaks, which is why the above formatting works (or should).
```

austinwiltshire commented on make_european.py:
```diff
@@ -0,0 +1,82 @@
+    day = int(date.group(2))
+    year = int(date.group(3))
+
+    assert 1 <= month <= 12, "Please enter a valid month."
```
also good assertion

austinwiltshire commented on make_european.py:
```diff
@@ -0,0 +1,82 @@
+
+    assert 1 <= month <= 12, "Please enter a valid month."
+
+    month31 = [1, 3, 5, 7, 8, 10, 12]
```
very detailed! :+1:

CoitThomas committed:
Add test modules for acronym.py and underscores.py.
Pull the function get_input out into its own module.

CoitThomas committed:
Implement a list comprehension in acronym module.
Removed get_input function from make_european module.
Simplify replace_space function in underscores module.

CoitThomas committed:
Remove excess assertions.
Implement 'Named Groups'.
Improve readability of the regular expressions in make_european module.

austinwiltshire commented:
looks good to me!
